src/osd/scrubber/osd_scrub_sched.cc

```diff
@@ -47,298 +48,152 @@ std::ostream& ScrubQueue::gen_prefix(std::ostream& out, std::string_view fn)
-      if (!filtr(jobref)) {
-	dout(20) << fmt::format(
-			" not eligible: {} @ {:s} ({:s},{:s})", jobref->pgid,
-			jobref->schedule.not_before,
-			jobref->schedule.scheduled_at, jobref->last_issue)
-		 << dendl;
-      }
+/**
+ * the set of all PGs named by the entries in the queue (but only those
+ * entries that satisfy the predicate)
+ */
+std::set<spg_t> ScrubQueue::get_pgs(const ScrubQueue::EntryPred& cond) const
+{
+  std::lock_guard lck{jobs_lock};
+  std::set<spg_t> pgs_w_matching_entries;
+  for (const auto& job : to_scrub) {
+    if (cond(*job)) {
+      pgs_w_matching_entries.insert(job->pgid);
     }
   }
-
-  return ripes;
+  return pgs_w_matching_entries;
 }
 
 
-Scrub::scrub_schedule_t ScrubQueue::adjust_target_time(
-  const sched_params_t& times) const
+#if !__clang__ || (__clang_major__ >= 16)
+void ScrubQueue::for_each_job(
+    std::function<void(const Scrub::ScrubJob&)> fn,
+    int max_jobs) const
 {
-  Scrub::scrub_schedule_t sched_n_dead{
-    times.proposed_time, times.proposed_time, times.proposed_time};
-
-  if (times.is_must == Scrub::must_scrub_t::not_mandatory) {
-    // unless explicitly requested, postpone the scrub with a random delay
-    double scrub_min_interval = times.min_interval > 0
-				  ? times.min_interval
-				  : conf()->osd_scrub_min_interval;
-    double scrub_max_interval = times.max_interval > 0
-				  ? times.max_interval
-				  : conf()->osd_scrub_max_interval;
-
-    sched_n_dead.scheduled_at += scrub_min_interval;
-    double r = rand() / (double)RAND_MAX;
-    sched_n_dead.scheduled_at +=
-      scrub_min_interval * conf()->osd_scrub_interval_randomize_ratio * r;
-
-    if (scrub_max_interval <= 0) {
-      sched_n_dead.deadline = utime_t{};
-    } else {
-      sched_n_dead.deadline += scrub_max_interval;
-    }
-    // note: no specific job can be named in the log message
-    dout(20) << fmt::format(
-		  "not-must. Was:{:s} {{min:{}/{} max:{}/{} ratio:{}}} "
-		  "Adjusted:{:s} ({:s})",
-		  times.proposed_time, fmt::group_digits(times.min_interval),
-		  fmt::group_digits(conf()->osd_scrub_min_interval),
-		  fmt::group_digits(times.max_interval),
-		  fmt::group_digits(conf()->osd_scrub_max_interval),
-		  conf()->osd_scrub_interval_randomize_ratio,
-		  sched_n_dead.scheduled_at, sched_n_dead.deadline)
-	     << dendl;
-  }
-  // else - no log needed. All relevant data will be logged by the caller
-  return sched_n_dead;
+  std::lock_guard lck(jobs_lock);
+  std::ranges::for_each(
+      to_scrub | std::views::take(max_jobs),
+      [fn](const auto& job) { fn(*job); });
 }
-
+#else
+// Clang does not support ranges for version < 16
+void ScrubQueue::for_each_job(
+    std::function<void(const Scrub::ScrubJob&)> fn,
+    int max_jobs) const
+{
+  std::lock_guard lck(jobs_lock);
+  std::for_each(
+      to_scrub.begin(),
+      to_scrub.begin() + std::min(max_jobs, static_cast<int>(to_scrub.size())),
+      [fn](const auto& job) { fn(*job); });
+}
```

Comment:
Let's not add two versions of this function.  Once our build environment fully supports ranges, feel free to switch it over.  Until then, I don't want it to be possible to fail to test the version users are actually using.

Reply:
removed

Reply:
done